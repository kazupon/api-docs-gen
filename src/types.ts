import type {
  ApiModel,
  ApiPackage,
  ApiItem
} from '@microsoft/api-extractor-model'

/**
 * Configuration
 */
export interface Config {
  /**
   * markdown link reference resolver
   */
  linkReferencer?: ReferenceResolver
  /**
   * markdown docs processor
   */
  processor: MarkdownProcessor
}

/**
 * Markdown reference resolver
 *
 * @param item the item of `api-extractor-model`
 * @param model the model of `api-extractor-model`
 *
 * @returns resolved reference path
 */
export type ReferenceResolver = (item: ApiItem, model: ApiModel) => string

/**
 * Markdown content
 */
export interface MarkdownContent {
  /**
   * markdown filename
   */
  filename: string
  /**
   * mkarkdown content
   */
  body: string
}

/**
 * Markdown docs processor
 *
 * @param model the api model of `api-extractor-model`
 * @param package the package of `api-extractor-model`
 * @param resolver the markdown reference resolver. if you're specfified at {@link Config}, it's passed, else it's not specified passed internal refenrece resolver.
 *
 * @returns markdown content
 */
export type MarkdownProcessor = (
  model: ApiModel,
  pkg: ApiPackage,
  resolver: ReferenceResolver
) => string | MarkdownContent[]
