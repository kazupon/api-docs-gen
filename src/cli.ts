import path from 'path'
import meow from 'meow'
import chalk from 'chalk'
import { debug as Debug } from 'debug'
import { generate } from './generator'

const debug = Debug('api-docs-gen:cli')

export const flags = {
  output: {
    type: 'string',
    alias: 'o',
    isRequired: false
  },
  config: {
    type: 'string',
    alias: 'c',
    isRequired: false
  }
} as const

const cli = meow(
  `
  Usage
    $ api-docs-gen <input>

  Options
    --config, -c configuration file
    --output, -o output dierectory that is markdown contents

  Examples
    $ api-docs-gen my-library-api.json
    $ api-docs-gen my-library-api.json --output ./docs
    $ api-docs-gen my-library-api.json --config docsgen.config.js
`,
  {
    flags
  }
)

// check input
const input = cli.input
if (!input.length) {
  console.error(chalk.red(`[api-docs-gen] not specified docs model`))
  process.exit(1)
}

// resolve output
const output =
  cli.flags.output != null ? path.resolve(cli.flags.output) : process.cwd()
debug(`output`, output)

// config
const config = cli.flags.config
debug(`config`, config)

// run
try {
  ;(async () => {
    await generate(input, output, config)
  })()
} catch (e) {
  console.error(chalk.red(`[api-docs-gen] ${e.message}`))
  process.exit(1)
}

process.on('uncaughtException', err => {
  console.error(chalk.red(`[api-docs-gen] Uncaught exception: ${err}\n`))
  process.exit(1)
})

process.on('unhandledRejection', (reason, p) => {
  console.error(
    chalk.red(`[api-docs-gen] Unhandled rejection at: ${p}, reason: ${reason}`)
  )
  process.exit(1)
})
