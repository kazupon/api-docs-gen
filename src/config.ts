import type { Resolver, Processor } from './types'

/**
 * Configration
 */
export interface Config {
  /**
   * markdown link reference resolver
   */
  resolver?: Resolver
  /**
   * processor
   */
  processor: Processor
}
