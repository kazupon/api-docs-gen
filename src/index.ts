export { getSafePathFromDisplayName, escapeText } from './utils'
export * from './builder'
export * from './config'
export {
  multi as multiProcessor,
  toc as tocProcessor,
  getCustomTags,
  getDocSectionContent
} from './processor'
export { multi as multiResolver, toc as tocResolver } from './resolver'
export { generate } from './generator'
