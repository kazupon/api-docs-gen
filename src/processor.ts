import {
  ApiItemKind,
  ApiModel,
  ApiPackage,
  ApiItem,
  ApiDeclaredItem,
  ApiParameterListMixin,
  IResolveDeclarationReferenceResult,
  ApiEnumMember
} from '@microsoft/api-extractor-model'
import {
  DocBlock,
  DocNodeKind,
  DocSection,
  DocPlainText,
  DocCodeSpan,
  IDocLinkTagParameters,
  DocDeclarationReference,
  DocFencedCode
} from '@microsoft/tsdoc'
import type { DeclarationReference } from '@microsoft/tsdoc/lib/beta/DeclarationReference'
import { debug as Debug } from 'debug'
import type {
  MarkdownContent,
  ReferenceResolver,
  GenerateStyle
} from './config'
import { ContentBuilder, createContentBuilder } from './builder'
import { escapeText } from './utils'

const debug = Debug('api-docs-gen:processor')

export function process(
  model: ApiModel,
  pkg: ApiPackage,
  style: GenerateStyle,
  resolver: ReferenceResolver
): string | MarkdownContent[] {
  // build
  function build(pkg: ApiPackage): Map<string, ContentBuilder> {
    const builders = new Map<string, ContentBuilder>()
    const items = pkg.members[0] ? pkg.members[0].members : []

    buildMarkdown(builders, items)
    return builders
  }

  // build markdown
  function buildMarkdown(
    builders: Map<string, ContentBuilder>,
    items: readonly ApiItem[]
  ): void {
    for (const item of items) {
      // console.log('buildMarkdown item', item)
      const { kind } = item
      let builder = builders.get(kind)
      if (!builder) {
        builder = createContentBuilder()
        builder.pushline(`# ${kind}`)
        builder.newline()
        builders.set(kind, builder)
      }

      switch (kind) {
        case ApiItemKind.Function:
          buildFunctionMarkdown(builder, item)
          break
        case ApiItemKind.Enum:
          buildEnumMarkdown(builder, item)
          break
        case ApiItemKind.Interface:
          buildInterfaceMarkdown(builder, item)
          break
        case ApiItemKind.Class:
          buildClassMarkdown(builder, item)
          break
        case ApiItemKind.TypeAlias:
          buildTypeAliasMarkdown(builder, item)
          break
        case ApiItemKind.Variable:
          buildVariableMarkdown(builder, item)
          break
        default:
          break
      }
    }
  }

  // build function markdown
  function buildFunctionMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // parameters
    const itemParam = item as ApiParameterListMixin
    if (itemParam.parameters) {
      builder.pushline(`### Parameters`)
      builder.newline()
      builder.pushline(`| Parameter | Type | Description |`)
      builder.pushline(`| --- | --- | --- |`)
      for (const p of itemParam.parameters) {
        builder.pushline(
          `| ${p.name} | ${p.parameterTypeExcerpt.text.trim()} | ${
            p.tsdocParamBlock && p.tsdocParamBlock.content
              ? getDocSectionContent(
                  model,
                  pkg,
                  // @ts-ignore TODO:
                  p.tsdocParamBlock.content,
                  item,
                  style,
                  resolver
                )
              : ''
          } |`
        )
      }
      builder.newline()
    }

    // returns
    if (docs.returnsBlock) {
      builder.pushline(`### Returns`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.returnsBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // throws
    // @ts-ignore TODO:
    const throws = getCustomTags(docs.customBlocks, '@throws')
    if (throws.length > 0) {
      builder.pushline(`### Throws`)
      builder.newline()
      for (const t of throws) {
        let text = `${getDocSectionContent(
          model,
          pkg,
          t.content,
          item,
          style,
          resolver
        )}`
        if (throws.length > 1) {
          text = `- ` + text
        }
        builder.pushline(text)
      }
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  // build enum markdown
  function buildEnumMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // members
    if (item.members) {
      builder.pushline(`### Members`)
      builder.newline()
      builder.pushline(`| Member | Value| Description |`)
      builder.pushline(`| --- | --- | --- |`)
      for (const m of item.members) {
        const memberEnum = m as ApiEnumMember
        const memberDeclared = m as ApiDeclaredItem
        builder.pushline(
          `| ${memberEnum.displayName} | ${
            memberEnum.initializerExcerpt.text
          } | ${
            memberDeclared.tsdocComment &&
            memberDeclared.tsdocComment.summarySection
              ? getDocSectionContent(
                  model,
                  pkg,
                  // @ts-ignore TODO:
                  memberDeclared.tsdocComment.summarySection,
                  item,
                  style,
                  resolver
                )
              : ''
          } |`
        )
      }
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  function buildMarkdownForClassinizable(
    builder: ContentBuilder,
    item: ApiItem,
    type: string
  ) {
    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // parameters
    const itemParam = item as ApiParameterListMixin
    if ((type === 'constrcutor' || type === 'method') && itemParam.parameters) {
      builder.pushline(`*Parameters*`)
      builder.newline()
      builder.pushline(`| Parameter | Type | Description |`)
      builder.pushline(`| --- | --- | --- |`)
      for (const p of itemParam.parameters) {
        builder.pushline(
          `| ${p.name} | ${p.parameterTypeExcerpt.text.trim()} | ${
            p.tsdocParamBlock && p.tsdocParamBlock.content
              ? getDocSectionContent(
                  model,
                  pkg,
                  // @ts-ignore TODO:
                  p.tsdocParamBlock.content,
                  item,
                  style,
                  resolver
                )
              : ''
          } |`
        )
      }
      builder.newline()
    }

    // returns
    if (type === 'method' && docs.returnsBlock) {
      builder.pushline(`### Returns`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.returnsBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // throws
    // @ts-ignore TODO:
    const throws = getCustomTags(docs.customBlocks, '@throws')
    if ((type === 'constructor' || type === 'method') && throws.length > 0) {
      builder.pushline(`### Throws`)
      builder.newline()
      for (const t of throws) {
        let text = `${getDocSectionContent(
          model,
          pkg,
          t.content,
          item,
          style,
          resolver
        )}`
        if (throws.length > 1) {
          text = `- ` + text
        }
        builder.pushline(text)
      }
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  // build interface markdown
  function buildInterfaceMarkdown(
    builder: ContentBuilder,
    item: ApiItem
  ): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()
    buildMarkdownForClassinizable(builder, item, 'interface')
    builder.newline()

    const methods = item.members.filter(m => m.kind === 'MethodSignature')
    builder.pushline(`### Methods`)
    builder.newline()
    for (const method of methods) {
      builder.pushline(`#### ${method.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, method, 'method')
    }
    builder.newline()

    const properties = item.members.filter(m => m.kind === 'PropertySignature')
    builder.pushline(`### Properties`)
    builder.newline()
    for (const property of properties) {
      builder.pushline(`#### ${property.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, property, 'property')
    }
    builder.newline()
  }

  // build class markdown
  function buildClassMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()
    buildMarkdownForClassinizable(builder, item, 'class')
    builder.newline()

    const [ctor] = item.members.filter(m => m.kind === 'Constructor')
    builder.pushline(`### Constructor`)
    builder.newline()
    buildMarkdownForClassinizable(builder, ctor, 'constructor')
    builder.newline()

    const methods = item.members.filter(m => m.kind === 'Method')
    builder.pushline(`### Methods`)
    builder.newline()
    for (const method of methods) {
      builder.pushline(`#### ${method.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, method, 'method')
    }
    builder.newline()

    const properties = item.members.filter(m => m.kind === 'Property')
    builder.pushline(`### Properties`)
    builder.newline()
    for (const property of properties) {
      builder.pushline(`#### ${property.displayName}`)
      builder.newline()
      buildMarkdownForClassinizable(builder, property, 'property')
    }
    builder.newline()
  }

  // build type alias markdown
  function buildTypeAliasMarkdown(
    builder: ContentBuilder,
    item: ApiItem
  ): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  // build variable markdown
  function buildVariableMarkdown(builder: ContentBuilder, item: ApiItem): void {
    builder.pushline(`## ${item.displayName}`)
    builder.newline()

    const itemDeclared = item as ApiDeclaredItem
    const docs = itemDeclared.tsdocComment
    if (!docs) {
      return
    }

    // summary
    if (docs.summarySection) {
      builder.pushline(
        // @ts-ignore
        getDocSectionContent(model, pkg, docs.summarySection, item, resolver)
      )
      builder.newline()
    }

    // signature
    if (itemDeclared.excerptTokens) {
      builder.pushline(`**Signature:**`)
      builder.pushline('```typescript')
      builder.pushline(
        itemDeclared.excerptTokens.map(token => token.text).join('')
      )
      builder.pushline('```')
      builder.newline()
    }

    // remarks
    if (docs.remarksBlock) {
      builder.pushline(`### Remarks`)
      builder.newline()
      builder.pushline(
        getDocSectionContent(
          model,
          pkg,
          // @ts-ignore TODO:
          docs.remarksBlock.content,
          item,
          style,
          resolver
        )
      )
      builder.newline()
    }

    // examples
    // @ts-ignore TODO:
    const examples = getCustomTags(docs.customBlocks, '@example')
    if (examples.length > 0) {
      builder.pushline(`### Examples`)
      builder.newline()
      let count = 1
      for (const e of examples) {
        if (examples.length > 1) {
          builder.pushline(`#### Example ${count}`)
        }
        builder.pushline(
          `${getDocSectionContent(
            model,
            pkg,
            e.content,
            item,
            style,
            resolver
          )}`
        )
        builder.newline()
        count++
      }
      builder.newline()
    }
  }

  function generate(builders: Map<string, ContentBuilder>): MarkdownContent[] {
    const contents = []
    for (const [kind, builder] of builders) {
      const content = builder.content
      contents.push({
        filename: `${kind.toLowerCase()}.md`,
        body: content
      })
    }
    return contents
  }

  // build markdown content!
  const builders = build(pkg)

  // generate markdown content!
  return generate(builders)
}

export function getDocSectionContent(
  model: ApiModel,
  pkg: ApiPackage,
  content: DocSection,
  contextItem: ApiItem,
  style: GenerateStyle,
  resolver: ReferenceResolver
): string {
  let ret = ''

  for (const n of content.nodes) {
    if (n.kind === DocNodeKind.Paragraph) {
      for (const nn of n.getChildNodes()) {
        if (nn.kind === DocNodeKind.PlainText) {
          ret += ((nn as unknown) as DocPlainText).text
        } else if (nn.kind === DocNodeKind.CodeSpan) {
          ret += `\`${((nn as unknown) as DocCodeSpan).code}\``
        } else if (nn.kind === DocNodeKind.LinkTag) {
          const {
            linkText,
            urlDestination,
            codeDestination
          } = (nn as unknown) as IDocLinkTagParameters
          if (codeDestination) {
            const result = resolveDeclarationReference(
              model,
              codeDestination,
              contextItem
            )
            if (result.resolvedApiItem) {
              const filepath = resolver(
                style,
                result.resolvedApiItem,
                model,
                pkg
              )
              if (linkText) {
                const encodedLinkText = escapeText(
                  linkText.replace(/\s+/g, ' ')
                )
                ret += `[${encodedLinkText}](${filepath})`
              } else {
                ret += `[${result.resolvedApiItem.displayName}](${filepath})`
              }
            }
          } else {
            if (linkText) {
              if (urlDestination) {
                const encodedLinkText = escapeText(
                  linkText.replace(/\s+/g, ' ')
                )
                ret += `[${encodedLinkText}](${urlDestination})`
              } else {
                ret += escapeText(linkText.replace(/\s+/g, ' '))
              }
            } else {
              if (urlDestination) {
                ret += urlDestination
              }
            }
          }
        }
      }
    } else if (n.kind === DocNodeKind.FencedCode) {
      const fenced = (n as unknown) as DocFencedCode
      ret += `\n`
      ret += `\`\`\`${fenced.language}\n`
      ret += fenced.code
      ret += `\`\`\``
    } else {
      debug(`ignore build doc section: ${n.kind}`)
    }
  }

  return ret
}

export function getCustomTags(
  customBlocks: readonly DocBlock[],
  tag: string
): DocBlock[] {
  return customBlocks.filter(x => x.blockTag.tagName === tag)
}

function resolveDeclarationReference(
  model: ApiModel,
  reference: DocDeclarationReference | DeclarationReference,
  item: ApiItem
): IResolveDeclarationReferenceResult {
  // @ts-ignore TODO:
  return model.resolveDeclarationReference(reference, item)
}
