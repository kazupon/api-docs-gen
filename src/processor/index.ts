export { process as multi } from './multi'
export { process as toc } from './toc'
export * from './utils'
