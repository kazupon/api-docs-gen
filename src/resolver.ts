import type { ApiItem, ApiModel } from '@microsoft/api-extractor-model'
import { getSafePathFromDisplayName } from './utils'

export function resolve(item: ApiItem, model: ApiModel): string {
  if (item.kind === 'Model') {
    return './index'
  }

  let baseName = ''
  for (const hierarchyItem of item.getHierarchy()) {
    const qualifiedName = getSafePathFromDisplayName(hierarchyItem.displayName)
    switch (hierarchyItem.kind) {
      case 'Model':
      case 'EntryPoint':
      case 'Package':
        break
      case 'Enum':
      case 'Function':
      case 'Variable':
      case 'TypeAlias':
        baseName += `${hierarchyItem.kind.toLowerCase()}#${qualifiedName}`
        break
      default:
        baseName += '.' + qualifiedName
    }
  }
  return './' + baseName
}
