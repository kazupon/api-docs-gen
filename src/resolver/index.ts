export { resolve as multi } from './multi'
export { resolve as toc } from './toc'
