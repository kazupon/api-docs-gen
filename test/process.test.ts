import path from 'path'
import { ApiModel } from '@microsoft/api-extractor-model'
import { resolve } from '../src/resolver'
import { process } from '../src/processor'
import { MarkdownContent } from '../src/types'

test('processor', () => {
  const apiModel = new ApiModel()
  const target = path.resolve(__dirname, './fixtures/my-library.api.json')
  const apiPackage = apiModel.loadPackage(target)
  const contents = process(apiModel, apiPackage, resolve) as MarkdownContent[]
  for (const content of contents) {
    expect(content.body).toMatchSnapshot(content.filename)
  }
})
