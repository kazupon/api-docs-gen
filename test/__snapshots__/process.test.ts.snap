// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processor: Class.md 1`] = `
"# Class

## Calculator

Calculator class

**Signature:**
\`\`\`typescript
export declare class Calculator implements Calculatable 
\`\`\`

### Remarks

This is remarks of Calculator class

### Examples


\`\`\`javascript
const c = new Calculator()
const v1 = c.add(1, 1)
const v2 = c.sub(1, 1)
\`\`\`



### Constructor

Conssutructor of usage

**Signature:**
\`\`\`typescript
constructor(type: string);
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 
| b | number | target 2 

#### sub

sub method

**Signature:**
\`\`\`typescript
sub(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 
| b | number | target 2 


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`

#### type

calcurator types 'simple'

**Signature:**
\`\`\`typescript
type: string;
\`\`\`


"
`;

exports[`processor: Enum.md 1`] = `
"# Enum

## ErrorCodes

Error Code

**Signature:**
\`\`\`typescript
export declare enum ErrorCodes 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| InvalidFormat | 1 | Invalid format 
| Succcess | 0 | Success 

## TokenChars

Token Characters

**Signature:**
\`\`\`typescript
export declare enum TokenChars 
\`\`\`

### Members

| Member | Value| Description |
| --- | --- | --- |
| Modulo | \\"%\\" | Modulo charactor 
| Pipe | \\"|\\" | Pipe charactor \`|\` 

### Remarks

This is remarks of Token Chararaceters

"
`;

exports[`processor: Function.md 1`] = `
"# Function

## add

add function : \`x\`

**Signature:**
\`\`\`typescript
export declare function add(a: number, b: number): number;
\`\`\`

### Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target \`1\` 
| b | number | target 2 

### Returns

 result as \`a\` + \`b\`

### Throws

- \`SyntaxError\` this is syntax error
- \`Error\` this is general error

### Remarks

This is add function remarks: [ErrorCodes](./enum#errorcodes). See [add](https://foo.bar.com/api/add).

### Examples

example of \`add\` function:
\`\`\`javascript
console.log(add(1, 1))
\`\`\`


"
`;

exports[`processor: Interface.md 1`] = `
"# Interface

## Calculatable

Calculatable interface

**Signature:**
\`\`\`typescript
export interface Calculatable 
\`\`\`


### Methods

#### add

add method

**Signature:**
\`\`\`typescript
add(a: number, b: number): number;
\`\`\`

*Parameters*

| Parameter | Type | Description |
| --- | --- | --- |
| a | number | target 1 
| b | number | target 2 


### Properties

#### PI

PI

**Signature:**
\`\`\`typescript
PI: number;
\`\`\`


"
`;

exports[`processor: TypeAlias.md 1`] = `
"# TypeAlias

## FallbackLocale

Falback Locale

**Signature:**
\`\`\`typescript
export declare type FallbackLocale = Locale | Locale[] | {
    [locale in string]: Locale[];
} | false;
\`\`\`

### Remarks

This is remarks of \`Fallback Locale\`

## Locale

Locale

**Signature:**
\`\`\`typescript
export declare type Locale = string;
\`\`\`

## LocaleMessage

Locale Message resources

**Signature:**
\`\`\`typescript
export declare type LocaleMessage = string | LocaleMessage[];
\`\`\`

"
`;

exports[`processor: Variable.md 1`] = `
"# Variable

## Config

Configrations 

**Signature:**
\`\`\`typescript
Config: {
    [name: string]: unknown;
}
\`\`\`

## VERSION

app version

**Signature:**
\`\`\`typescript
VERSION = \\"1.0.0\\"
\`\`\`

### Remarks

you can lookup application version that is semver format.

### Examples


\`\`\`javascript
console.log(VERSION)
\`\`\`


"
`;
